libs/json/decoder.go

```diff
@@ -0,0 +1,278 @@
+package json
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+// Unmarshal unmarshals JSON into the given value, using Amino-compatible JSON encoding (strings
+// for 64-bit numbers, and type wrappers for registered types).
+func Unmarshal(bz []byte, v interface{}) error {
+	return decode(bz, v)
+}
+
+func decode(bz []byte, v interface{}) error {
+	if len(bz) == 0 {
+		return errors.New("cannot decode empty bytes")
+	}
+
+	rv := reflect.ValueOf(v)
+	if rv.Kind() != reflect.Ptr {
+		return errors.New("must decode into a pointer")
+	}
+	rv = rv.Elem()
+
+	// If this is a registered type, defer to interface decoder regardless of whether the input is
+	// an interface or a bare value. This retains Amino's behavior, but is inconsistent with
+	// behavior in structs where an interface field will get the type wrapper while a bare value
+	// field will not.
+	if typeRegistry.name(rv.Type()) != "" {
+		return decodeReflectInterface(bz, rv)
+	}
+
+	return decodeReflect(bz, rv)
+}
+
+func decodeReflect(bz []byte, rv reflect.Value) error {
+	if !rv.CanAddr() {
+		return errors.New("value is not addressable")
+	}
+
+	// Handle null for slices, interfaces, and pointers
+	if bytes.Equal(bz, []byte("null")) {
+		rv.Set(reflect.Zero(rv.Type()))
+		return nil
+	}
+
+	// Dereference-and-construct pointers, to handle nested pointers.
+	for rv.Kind() == reflect.Ptr {
+		if rv.IsNil() {
+			rv.Set(reflect.New(rv.Type().Elem()))
+		}
+		rv = rv.Elem()
+	}
+
+	// Times must be UTC and end with Z
+	if rv.Type() == timeType {
+		switch {
+		case len(bz) < 2 || bz[0] != '"' || bz[len(bz)-1] != '"':
+			return fmt.Errorf("JSON time must be an RFC3339 string, but got %s", bz)
+		case bz[len(bz)-2] != 'Z':
```

Comment:
don't we want to check both cases here?

Reply:
What other case?

Reply:
Not sure I follow. Both branches are checked? Do you mean combining them into a single check with a common error message?

Reply:
sorry, it's just a confusing style of checking for errors...

i'd much prefer 

if X {
  return E
}

if Y {
  return E
}